ops/terraform/services/pipeline/modules/bfd_pipeline_manager/lambda_src/pipeline_manager.py

```diff
@@ -0,0 +1,352 @@
+    )
+    # Returns the file names of all text files within the "done" folder for the current bucket
+    finished_rifs = [
+        str(object.key).removeprefix(done_prefix)
+        for object in etl_bucket.objects.filter(Prefix=done_prefix)
+        if str(object.key).endswith(".txt") or str(object.key).endswith(".csv")
+    ]
+
+    # We check for all RIFs _except_ beneficiary history as beneficiary history is a RIF type not
+    # expected to exist in CCW-provided loads -- it only appears in synthetic loads
+    rif_types_to_check = [e for e in RifFileType if e != RifFileType.BENEFICIARY_HISTORY]
+
+    # Check, for each rif file type, if any finished rif has the corresponding rif type prefix.
+    # Essentially, this ensures that all non-optional (excluding beneficiary history) RIF types have
+    # been loaded and exist in the Done/ folder
+    return all(
+        any(rif_type.value in rif_file_name.lower() for rif_file_name in finished_rifs)
+        for rif_type in rif_types_to_check
+    )
+
+
+def _is_incoming_folder_empty(load_type: PipelineLoadType, group_timestamp: str) -> bool:
+    incoming_key_prefix = (
+        "/".join(
+            filter(
+                None,
+                [load_type.capitalize(), PipelineDataStatus.INCOMING.capitalize(), group_timestamp],
+            )
+        )
+        + "/"
+    )
+    incoming_objects = list(etl_bucket.objects.filter(Prefix=incoming_key_prefix))
+
+    return len(incoming_objects) == 0
+
+
+def handler(event: Any, context: Any):
+    if not all(
+        [
+            REGION,
+            BFD_ENVIRONMENT,
+            DEPLOYED_GIT_BRANCH,
+            JENKINS_TARGET_JOB_NAME,
+            JENKINS_JOB_RUNNER_QUEUE,
+            ONGOING_LOAD_QUEUE,
+            ETL_BUCKET_ID,
+        ]
+    ):
+        print("Not all necessary environment variables were defined, exiting...")
+        return
+
+    try:
+        record: dict[str, Any] = event["Records"][0]
+    except KeyError as exc:
+        print(f"The incoming event was invalid: {exc}")
+        return
+    except IndexError:
+        print("Invalid event notification, no records found")
+        return
+
+    try:
+        file_key: str = record["s3"]["object"]["key"]
+    except KeyError as exc:
+        print(f"No bucket file found in event notification: {exc}")
+        return
+
+    decoded_file_key = unquote(file_key)
+    status_group_str = "|".join([e.value for e in PipelineDataStatus])
+    rif_types_group_str = "|".join([e.value for e in RifFileType])
+    # The incoming file's key should match an expected format, as follows:
+    # "<Synthetic/>/<Incoming/Done>/<ISO date format>/<file name>".
+    if match := re.search(
+        pattern=(
+            rf"^({PipelineLoadType.SYNTHETIC.capitalize()}){{0,1}}/{{0,1}}"
+            rf"({status_group_str})/"
+            rf"([\d\-:TZ]+)/"
+            rf".*({rif_types_group_str}).*$"
+        ),
+        string=decoded_file_key,
+        flags=re.IGNORECASE,
```

Comment:
Should we do this if we're also purposefully capitalizing values like load type, and then doing case-sensitive comparisons?

Reply:
I don't _think_ it's necessary to use ignore case here, but I think it helps cover any case where a RIF file is uploaded with a mixed case name (i.e. `Bene_...` vs. `bene_....`).

I replaced the `PipelineLoadType.SYNTHETIC.capitalize()` with `PipelineLoadType.SYNTHETIC.value` in the regex pattern as, like you are suggesting, the casing of the pattern doesn't matter. That was a bit of an oversight; there are other places in the code where I explicitly `capitalize()` some string values (that are usually already capitalized) just to make things a little more explicit, but this spot is entirely unnecessary.